### PR TITLE
Prepare test and pull/push enhancements

### PR DESCRIPTION
**Pull request by jscotka**
it fixes: https://github.com/psss/tmt/issues/587

### CODE REVIEW

**Comment by lukaszachy** [comment]
Looks reasonable, could you make it work for container as well?

Finally one can use script file in prepare.

**Comment by jscotka** [comment]
> Looks reasonable, could you make it work for container as well?
> 
> Finally one can use script file in prepare.
> 
> Test plan used for experiment:
> 
> ```
> $ ls -1 | xargs head
> ==> p.fmf <==
> discover:
>  tests:
>  - name: fake
>    test: true
> prepare:
>  how: shell
>  script: bash prepare_script
> execute:
>  how: tmt
> 
> ==> prepare_script <==
> echo DONE
> ```
> 
> ```
> prepare
>         workdir: /var/tmp/tmt/run-324/p/prepare
>         Using the 'PrepareInstall' plugin for the 'install' method.
>         how: shell
>         order: 50
>         overview: 1 script found
>         script: bash prepare_script
>             Execute command 'mkdir -p /var/tmp/tmt/run-324/p/prepare' on guest '192.168.122.25'.
>             Run command 'ssh -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null root@192.168.122.25 mkdir -p /var/tmp/tmt/run-324/p/prepare'.
>             err: Warning: Permanently added '192.168.122.25' (ECDSA) to the list of known hosts.
>             Copy files to guest '192.168.122.25'.
>             Run command 'rsync -rz --links --safe-links -e "ssh -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null" /tmp/a//* root@192.168.122.25:/var/tmp/tmt/run-324/p/prepare'.
>             err: Warning: Permanently added '192.168.122.25' (ECDSA) to the list of known hosts.
>             Execute command 'bash prepare_script' on guest '192.168.122.25'.
>             Run command 'ssh -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null root@192.168.122.25 cd '/var/tmp/tmt/run-324/p/prepare'; bash prepare_script'.
>             err: Warning: Permanently added '192.168.122.25' (ECDSA) to the list of known hosts.
>             out: DONE
> ```

@lukaszachy Thank you very much, fixed also for podman container.

**Review by psss — CHANGES_REQUESTED** [review]
Thanks for the proposal. The test looks nice but for the fix itself I would suggest we take a more systematic approach. Let's sumarize what we've discussed already:

 * #396 ... sharing git repos across plans
 * #423 ... finish and execute should run in the same folder
 * #497 ... cannot access workdir in prepare
 * #587 ... prepare cannot cd to non-existent directory

The first two issues outline the solution by sharing the working `tree` using `git worktree` feature which seems to me as the best way. Together with executing all `prepare` and `finish` step scripts in the this directory we should have the issue covered.

I'm not yet completely sure about how to exactly handle `discover --how shell` and `execute --script`. As for now these run under the `discover/<name>/tests` directory. And this seems to be working fine. Probably no changes need (except that we could save some more disk space if we used `git worktree` here as well).

Let's finalize the discussion about the design first before we dive into the implementation.

**Comment by psss** [comment]
Summary from the discussion:

 * `/tmp/tmt/run-001/tree` will contain complete git repo from which tests are run
 * `/tmp/tmt/run-001/plan/tree` will be a detached git worktree of above

The following steps will have cwd set to `plans/tree`:

 * `prepare`
 * `finish`

Also `login` will default to jump to the `tree` directory for whichever step it is used.

As the next steps we can improve:

 * `discover` to use symlink if `url` and `ref` are not provided
 * sync only `discover` and `execute` workdirs to guest

**Comment by psss** [comment]
So the next steps have been outlined and @thrix will be working on the shared `tree` directory implemenation. What about this pull request? I'd suggest to limit it to the new test coverage (which will be useful) and optionally to the extended functionality of `push()` and `pull()`. What do you think, @jscotka?

**Comment by jscotka** [comment]
> So the next steps have been outlined and @thrix will be working on the shared `tree` directory implemenation. What about this pull request? I'd suggest to limit it to the new test coverage (which will be useful) and optionally to the extended functionality of `push()` and `pull()`. What do you think, @jscotka?

Sure, makes sense, if you see functionality of `copy_from` and `copy_to` as useful then yes. I needed it to copy what I wanted to guest efficiently. I think it could be also sometimes more efficient than this pull/push, in some cases.

This test coverage yes, it is useful, sure. After it will be implemented by miro (with possible extension to all provisioners).

But now I do not know. how to handle it @psss ? if miro will continue with this PR and use appropriate changes, or of I should change this PR and let there just exclusive code, what will miro does not touch, or if this PR will be just base of code for another PR's, and after others will be finished, we  can close this as well.

**Comment by psss** [comment]
I would leave this pull request just for the new test and extended `push()` and `pull()` methods. Here's my suggestion: Instead of introducing new methods, let's extend the existing ones with optional parameters `src` and `dest`. If only `src` is provided, that directory would be synced to the same path on guest (useful for example for syncing `discover` step only) and if `dest` is provided, syncing to the given location. What do you think?

**Comment by jscotka** [comment]
> I would leave this pull request just for the new test and extended `push()` and `pull()` methods. Here's my suggestion: Instead of introducing new methods, let's extend the existing ones with optional parameters `src` and `dest`. If only `src` is provided, that directory would be synced to the same path on guest (useful for example for syncing `discover` step only) and if `dest` is provided, syncing to the given location. What do you think?

Make sense.

Just what I had issue is with default params for rsync (especially `-R` what convert to src to relative path, and changing this, will change current behaviour if I remove this and for me, usage of this is little bit strange will lead to path on guest like `/dest/path/home/jscotka/git/tmt/...` what is strange at all for me.)

**Comment by psss** [comment]
What about also adding a new `options` parameter which would default to the current behavior to keep it backward compatible?

**Comment by jscotka** [comment]
> What about also adding a new `options` parameter which would default to the current behavior to keep it backward compatible?

yes, this is good compromise

**Comment by psss** [comment]
Good, agreed then.

**Comment by jscotka** [comment]
@psss I've adapted the mentioned chanes. I hope it is what you wanted :-)

File: tmt/steps/provision/podman.py [hunk 1]
```diff
@@ -153,7 +153,10 @@ def ansible(self, playbook):
 
     def podman(self, command, **kwargs):
         """ Run given command via podman """
-        return self.run(['podman'] + command, shell=False, **kwargs)
+        if isinstance(command, str):
+            return self.run('podman ' + command, shell=True, **kwargs)
+        else:
+            return self.run(['podman'] + command, shell=False, **kwargs)
```
Comment (psss) [inline tmt/steps/provision/podman.py:159]: What is the motivation for this change? The `shell=False` approach should be a bit safer.

**Review by psss — COMMENTED** [review]
Thanks for the improvements. I've added a couple adjustments:

 * finalize the test (including the beakerlib code, so that it's prepared for future extensions)
 * use long options (sorry for the original proposal, while reviewing the code I realized it's (almost) always better to use the full name)
 * apply the default options even when `source` is provided (will be useful once we sync selected step workdirs only)

Plus one question regarding the `podman` plugin. Please, have a look. Thanks.

File: tmt/steps/provision/podman.py (hunk 1 above)
Comment (jscotka) [reply tmt/steps/provision/podman.py:159]: I've let there this code, because in original source i've used it to implement behaviour for pull/push files to container and shell were easier to pass options, than try to parse tem somehow. This part can be removed.

**Comment by jscotka** [comment]
> Thanks for the improvements. I've added a couple adjustments:
> 
> * finalize the test (including the beakerlib code, so that it's prepared for future extensions)
> * use long options (sorry for the original proposal, while reviewing the code I realized it's (almost) always better to use the full name)
> * apply the default options even when `source` is provided (will be useful once we sync selected step workdirs only)
> 
> Plus one question regarding the `podman` plugin. Please, have a look. Thanks.

Thanks for changes. I agree with your changes.

**Review by psss — APPROVED** [review]
Ok, good. Squashed & rebased.
